crowbar_framework/app/models/crowbar_service.rb

```diff
@@ -176,6 +176,49 @@ def create_proposal
     base
   end
 
+  def prepare_nodes_for_os_upgrade
+    nodes_to_upgrade = []
+    all_nodes = NodeObject.all
+
+    all_nodes.each do |node|
+      next if node.admin? or node[:platform] == "windows"
+      node.set_state("os_upgrading")
+    end
+
+    # wait for the pxe_config to be updated, then reboot the nodes
+    discovery_dir = "#{NodeObject.admin_node[:provisioner][:root]}/discovery/"
+    pxecfg_subdir = "bios/pxelinux.cfg"
+
+    all_nodes.each do |node|
+      next if node.admin? or node["platform"] == "windows"
+
+      boot_ip_hex = node["crowbar"]["boot_ip_hex"]
+      node_arch = node["kernel"]["machine"] rescue "x86_64"
+
+      pxe_conf = "#{discovery_dir}/#{node_arch}/#{pxecfg_subdir}/#{boot_ip_hex}"
+      ready_for_reboot = false
+
+      while Time.now.to_i < Time.now.to_i + 120 && !ready_for_reboot do
```

Comment:
Style/WhileUntilDo: Do not use `do` with multi-line `while`. (https://github.com/bbatsov/ruby-style-guide#no-multiline-while-do)